File: 백준/Platinum/2150. Strongly Connected Component/Strongly Connected Component.java
import java.io.*;
import java.util.*;

public class Main {
    static int V, E, id;
    static int[] p;
    static boolean[] finished;
    static List<Integer>[] graph;
    static List<List<Integer>> scc;
    static Deque<Integer> stack;

    static int dfs(int cur) {
        p[cur] = ++id;
        stack.push(cur);

        int parent = p[cur];
        for(int nxt : graph[cur]) {
            if(p[nxt] == 0) {
                parent = Math.min(parent, dfs(nxt));
            } else if(!finished[nxt]) {
                parent = Math.min(parent, p[nxt]);
            }
        }

        if(parent == p[cur]) {
            List<Integer> ele = new ArrayList<>();
            while(!stack.isEmpty()) {
                int nxt = stack.pop();
                finished[nxt] = true;
                ele.add(nxt);
                if(cur == nxt) break;
            }
            Collections.sort(ele);
            scc.add(ele);
        }

        return parent;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;

        st = new StringTokenizer(br.readLine(), " ");
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        id = 0;
        p = new int[V + 1];
        finished = new boolean[V + 1];
        graph = new List[V + 1];
        scc = new ArrayList<>();
        stack = new ArrayDeque<>();
        for(int i = 1; i <= V; i++) {
            graph[i] = new ArrayList<>();
        }

        for(int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            graph[A].add(B);
        }

        for(int i = 1; i <= V; i++) {
            if(p[i] == 0) dfs(i);
        }

        StringBuilder sb = new StringBuilder();
        sb.append(scc.size()).append('\n');
        Collections.sort(scc, (o1, o2) -> o1.get(0) - o2.get(0));
        for(List<Integer> ele : scc) {
            for(int i : ele) {
                sb.append(i).append(' ');
            }
            sb.append(-1).append('\n');
        }

        System.out.print(sb);
        br.close();
    }
}

File: 백준/Platinum/2150. Strongly Connected Component/Strongly Connected Component.cc
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <cstring>

using namespace std;
const int sz = 10'005;

int v, e, id;
int p[sz];
bool finished[sz];
vector<int> g[sz], scc[sz];
stack<int> stk;
int scc_cnt;

int dfs(int cur) {
    p[cur] = ++id;
    stk.push(cur);
    int par = p[cur];

    for (int nxt : g[cur]) {
        if (p[nxt] == 0) par = min(par, dfs(nxt));
        else if (!finished[nxt]) par = min(par, p[nxt]);
    }
    
    if (par == p[cur]) {
        while (1) {
            int t = stk.top(); stk.pop();
            finished[t] = 1;
            scc[scc_cnt].emplace_back(t);
            if (t == cur) break;
        }
        sort(scc[scc_cnt].begin(), scc[scc_cnt].end());
        ++scc_cnt;
    }

    return par;
}

int comp(vector<int> a, vector<int> b) {
    return a[0] < b[0];
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    cin >> v >> e;
    for (int i = 0; i < e; ++i) {
        int a, b;
        cin >> a >> b;
        g[a].emplace_back(b);
    }

    for (int i = 1; i <= v; ++i) {
        if (p[i] == 0) dfs(i);
    }

    vector<vector<int>> res(scc, scc + scc_cnt);
    sort(res.begin(), res.end(), comp);

    cout << scc_cnt << '\n';
    for (auto& s : res) {
        for (int x : s) cout << x << ' ';
        cout << -1 << '\n';
    }
    return 0;
}
